server/pulp/plugins/migration/standard_storage_path.py

```diff
@@ -0,0 +1,374 @@
+        :type key_fields: tuple
+        :param join_leaf: Join the existing path leaf (filename) to the new path.
+        :type join_leaf: bool
+        """
+        self.collection = collection
+        self.key_fields = key_fields
+        self.join_leaf = join_leaf
+
+    @property
+    def base_fields(self):
+        """
+        Unit fields to be included in the query.
+
+        :return: The base fields.
+        :rtype: tuple
+        """
+        return (
+            '_storage_path',
+            '_content_type_id'
+        )
+
+    def _new_path(self, unit):
+        """
+        Determine the *new* storage path for the specified unit.
+
+        :param unit: A unit being migrated.
+        :type unit: Unit
+        :return: The new storage path.
+        :rtype: str
+        """
+        digest = unit.key_digest()
+        path = os.path.join(
+            os.path.join(Migration.content_dir(), 'units'),
+            unit.type_id,
+            digest[0:2],
+            digest[2:])
+        if self.join_leaf:
+            path = os.path.join(path, os.path.basename(unit.storage_path))
+        return path
+
+    def __iter__(self):
+        """
+        Get an iterable of units planned to be migrated.
+
+        :return: Unit planned to be migrated.
+        :rtype: generator
+        :
+        """
+        fields = dict(
+            (k, True) for k in chain(self.key_fields, self.base_fields)
+        )
+        for document in self.collection.find(projection=fields):
+            unit = Unit(self, document)
+            unit.new_path = self._new_path(unit)
+            if not unit.needs_migration():
+                continue
+            yield unit
+
+
+class Migration(object):
+    """
+    Migration of unit `storage_path` to the standardized path
+    introduced in Pulp 2.8.
+      1. Execute each plan.
+      2. Prune content tree of empty directories.
+    """
+
+    @staticmethod
+    def storage_dir():
+        """
+        The root storage path.
+
+        :return: The root storage path.
+        :rtype: str
+        """
+        return config.get('server', 'storage_dir')
+
+    @staticmethod
+    def content_dir():
+        """
+        The root content storage path.
+
+        :return: The root content storage path.
+        :rtype: str
+        """
+        return os.path.join(Migration.storage_dir(), 'content')
+
+    @staticmethod
+    def publish_dir():
+        """
+        The root published path.
+
+        :return: The root storage path.
+        :rtype: str
+        """
+        return os.path.join(Migration.storage_dir(), 'published')
+
+    @staticmethod
+    def _prune():
+        """
+        Delete empty directories in the content storage tree.
+        The migration will create empty directories.
+        """
+        root = Migration.content_dir()
+        for path, directories, files in os.walk(root, topdown=False):
+            if not os.listdir(path):
```

Comment:
Oh, wait, now I remember.  The walk is yielding a snapshot of the tree.  During the bottom-up walk of the tree - if a leaf directory is deleted, it still show up in the `directories` list on the parent even though we just deleted it and the parent is now empty.